### DISCUSSION TITLE
Communicate over UART to four HC-05 with micropython

### DISCUSSION THREAD

**Discussion by leubanse** [micropython/micropython#11506] [RP2040 / Pico] [answered]
Hi together,

is it possible to communicate over UART to four HC-05 Bluetooth devices? I know, I have only two physical UARTs (0 and 1), but there are 5 pin pairs, which provide connection to TX and RX. 

It works, that I can send from RP2040 to bluetooth and it will be shown on the Android app "Serial Bluetooth Terminal". 

But when I try to send from Android app over HC-05 to RP2040, I can do it only once by UART0 or UART1. If I "re-initialize" UART with waiting 15 seconds after receiving, I can't receive the message.

The first UART1 and the first UART0 works but the second UART0 doesn't work and only machine.reset() can make it work again. Only stop the code, doesn't work. I'm using Thonny for print output and to run the code.

Version of firmware: rp2-pico-20230426-v1.20.0.uf2

Sample code:

```
import time
import machine

uart1 = None

def setUartPlayer(player: int):
  global uart1
  if player == 1:
    uart1 = machine.UART(1, baudrate=9600, tx=machine.Pin(4), rx=machine.Pin(5))
  elif player == 2:
    uart1 = machine.UART(0, baudrate=9600, tx=machine.Pin(16), rx=machine.Pin(17))
  elif player == 3:
    uart1 = machine.UART(0, baudrate=9600, tx=machine.Pin(12), rx=machine.Pin(13))
  elif player == 4:
    uart1 = machine.UART(1, baudrate=9600, tx=machine.Pin(8), rx=machine.Pin(9))

def communicate(player: int):
  global uart1
  setUartPlayer(player)
  print(uart1)
  uart1.write('test\n')
  while True:
    if uart1.any() > 0:
      data = uart1.read()
      print(data)
      return

print('go')
communicate(1)
time.sleep(15)
print('go')
communicate(2)
time.sleep(15)
print('go')
communicate(3)
time.sleep(15)
print('go')
communicate(4)
time.sleep(15)
```

Thanks for you help and regards,
Jens

**Comment by jimmo** [comment]
@leubanse  I wondered if maybe something is going wrong with re-initialising the UART, but I just tried this:

```
>>> import machine
>>> uart1 = machine.UART(1, baudrate=9600, tx=machine.Pin(4), rx=machine.Pin(5))
>>> uart1.write('hi')
2
```
(Send "there" from the other end of the UART)
```
>>> uart1.read()
b'there'
```
(Switch pins over)
```
>>> uart1 = machine.UART(1, baudrate=9600, tx=machine.Pin(8), rx=machine.Pin(9))
>>> uart1.write('hi')
2
```
(Send "there" from the other end of the UART)
```
>>> uart1.read()
b'there'
>>> uart1.read()
```

So I think this should work.

> but the second UART0 doesn't work and only machine.reset() can make it work again

When this happens, what do you see? Does the read()/write() fail, hang, return nothing?

**Comment by leubanse** [comment]
What I observed is, that "test" will be received by both UART0, what is strange, because the pins were changed. I will try a simple example today evening, without functions, only call the UARTs after each other. Let's see.

**Reply by leubanse** [reply]
Here is a simple example, which doesn't work as expected. The second "test" which should be received by `machine.UART(0, baudrate=9600, tx=machine.Pin(12), rx=machine.Pin(13))` will be received also by `machine.UART(0, baudrate=9600, tx=machine.Pin(16), rx=machine.Pin(17))`. It will be received by both UARTs and after "re-init" the UART, it is not possible anymore to send something back to UART.

```
import utime
import machine

uart0 = machine.UART(0, baudrate=9600, tx=machine.Pin(16), rx=machine.Pin(17))
print(uart0)
uart0.write('test\n')
prvMills = utime.ticks_ms()
while (utime.ticks_ms() - prvMills) < 15000:
  if uart0.any():
    data = uart0.read()
    print(data)

uart0 = machine.UART(0, baudrate=9600, tx=machine.Pin(12), rx=machine.Pin(13))
print(uart0)
uart0.write('test\n')
prvMills = utime.ticks_ms()
while (utime.ticks_ms() - prvMills) < 15000:
  if uart0.any():
    data = uart0.read()
    print(data)
```

**Reply by robert-hh** [reply, marked as answer]
You could try to reconfigure the UART pins after use as machine.Pin objects, which changes it's mode. e.g.
```
p17=machine.pin(17, Pin.IN)
p16=machine.pin(16, Pin.IN)
```
etc.

The assignment is not required

**Reply by leubanse** [reply]
Many thanks Robert, this is the solution, if I do a "reset" of the pins, it works :)

Full working example:

```
import utime
import machine

def resetPins():
  machine.Pin(12, machine.Pin.OUT)
  machine.Pin(13, machine.Pin.OUT)
  machine.Pin(16, machine.Pin.OUT)
  machine.Pin(17, machine.Pin.OUT)

resetPins()

uart0 = machine.UART(0, baudrate=9600, tx=machine.Pin(16), rx=machine.Pin(17))
print(uart0)
uart0.write('test\n')
prvMills = utime.ticks_ms()
while (utime.ticks_ms() - prvMills) < 20000:
  if uart0.any():
    data = uart0.read()
    print(data)

resetPins()

uart0 = machine.UART(0, baudrate=9600, tx=machine.Pin(12), rx=machine.Pin(13))
print(uart0)
uart0.write('test\n')
prvMills = utime.ticks_ms()
while (utime.ticks_ms() - prvMills) < 20000:
  if uart0.any():
    data = uart0.read()
    print(data)
```

**Reply by corlovsky** [reply]
This fixed it for me too!
